src/composeappmanager.cc

```diff
@@ -145,25 +227,274 @@ data::InstallationResult ComposeAppManager::install(const Uptane::Target &target
+  if (received_manifest_hash != digest.hash()) {
+    throw std::runtime_error(
+        "Hash of received App manifest and the hash specified in Target"
+        "do not match: " +
+        received_manifest_hash + " != " + digest.hash());
+  }
+
+  LOG_TRACE << "Received App manifest: \n" << manifest_resp.getJson();
+  return ComposeApp::Manifest{manifest_resp.getJson()};
+  ;
+}
+
+struct DownloadCtx {
+  DownloadCtx(std::ostream& out_stream_in, MultiPartHasher& hasher_in, std::size_t expected_size_in)
+      : out_stream{out_stream_in}, hasher{hasher_in}, expected_size{expected_size_in} {}
+
+  std::ostream& out_stream;
+  MultiPartHasher& hasher;
+  std::size_t expected_size;
+
+  std::size_t written_size{0};
+
+  std::size_t write(const char* data, std::size_t size) {
+    assert(data);
+
+    if ((written_size + size) > expected_size) {
+      LOG_ERROR << "!!! Received data size exceeds the expected size: " << (written_size + size)
+                << " != " << expected_size;
+      return (size + 1);  // returning value that is not equal to received data size will make curl fail
+    }
+
+    if (!out_stream.good()) {
+      LOG_ERROR << "Output stream is at a bad state: " << out_stream.rdstate();
+      return (size + 1);  // returning value that is not equal to received data size will make curl fail
+    }
+
+    auto start_pos = out_stream.tellp();
+    out_stream.write(data, size);
+    auto end_pos = out_stream.tellp();
+
+    written_size += (end_pos - start_pos);
+    hasher.update(reinterpret_cast<const unsigned char*>(data), size);
+    return (end_pos - start_pos);
+  }
+};
+
+static size_t DownloadHandler(char* data, size_t buf_size, size_t buf_numb, void* user_ctx) {
+  assert(user_ctx);
+
+  auto download_ctx = reinterpret_cast<DownloadCtx*>(user_ctx);
+  return download_ctx->write(data, (buf_size * buf_numb));
+}
+
+void RegistryClient::downloadBlob(const std::string& repo, const HashedDigest& digest,
+                                  const boost::filesystem::path& file_path, size_t expected_size) const {
+  auto compose_app_blob_url{conf_.repo_base_url + repo + "/blobs/" + digest()};
+
+  LOG_DEBUG << "Downloading App blob: " << compose_app_blob_url;
+
+  std::vector<std::string> registry_repo_request_headers{getBearerAuthHeader(repo)};
+  auto registry_repo_client{http_client_factory_(&registry_repo_request_headers)};
+
+  std::ofstream output_file{file_path.string(), std::ios_base::out | std::ios_base::binary};
+  if (!output_file.is_open()) {
+    throw std::runtime_error("Failed to open a file: " + file_path.string());
+  }
+  MultiPartSHA256Hasher hasher;
+
+  DownloadCtx download_ctx{output_file, hasher, expected_size};
+  auto get_blob_resp = registry_repo_client->download(compose_app_blob_url, DownloadHandler, nullptr, &download_ctx, 0);
+
+  if (!get_blob_resp.isOk()) {
+    throw std::runtime_error("Failed to download App blob: " + get_blob_resp.getStatusStr());
+  }
+
+  output_file.close();
```

Comment:
this is needed because the file normally won't  get closed until the destructor is called, right?

Reply:
yes.